segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StreamSegmentContainer.java

```diff
@@ -370,6 +372,27 @@ public int getId() {
     }
 
     @Override
+    public CompletableFuture<Long> truncateStreamSegment(String streamSegmentName, long offset, Duration timeout) {
+        ensureRunning();
+
+        logRequest("truncateStreamSegment", streamSegmentName);
+        this.metrics.truncate();
+        TimeoutTimer timer = new TimeoutTimer(timeout);
+        AtomicLong segmentId = new AtomicLong();
+        return this.segmentMapper
+                .getOrAssignStreamSegmentId(streamSegmentName, timer.getRemaining(),
+                        streamSegmentId -> {
+                            segmentId.set(streamSegmentId);
+                            StreamSegmentTruncateOperation op = new StreamSegmentTruncateOperation(streamSegmentId, offset);
+                            return this.durableLog.add(op, timer.getRemaining());
```

Comment:
Here this is returning a CompletableFuture<Void> from another method but setting segmentId to pass it to the thenApply block below. Ideally we would chain a future that returned segmentId on the result of durableLog.add

Reply:
`durableLog.add()` returns a `CompletableFuture<Void>` which completes when the operation has been processed. There are 10 types of operations, most of which do not deal with Segment Ids. As such, I can't have it return a SegmentId, so I need to extract it from the operation after its completion.

Reply:
I was not suggesting to change add() but rather to take it's result and compose it with a completed future of the segmentId and return that.